dataflow/src/org/checkerframework/dataflow/util/PurityChecker.java

```diff
@@ -357,6 +388,10 @@ public PurityResult visitMethodInvocation(MethodInvocationTree node, PurityResul
                     p.addNotDetReason(node, reason);
                 } else if (!seFree) {
                     p.addNotSeFreeReason(node, reason);
+                } else if (!sdet) {
```

Comment:
Are you sure that the sdet and mdet branches can be reached? Is the det branch reached first?
How about removing Deterministic for good and properly adapting this logic?